File: 1268-search-suggestions-system/1268-search-suggestions-system.go
func suggestedProducts(products []string, searchWord string) [][]string {
    
    sort.Slice(products, func(i, j int) bool {
        return products[i] < products[j]
    })
    
    result := [][]string{}
    
    left := 0
    right := len(products) - 1
    
    for i:=0; i < len(searchWord); i++ {
        strs := []string{}
        c := searchWord[i]
        
        for left <= right && (len(products[left]) <= i || products[left][i] < c) {
            left++
        }
        for left <= right && (len(products[right]) <= i || products[right][i] > c) {
            right --
        }
        
        for j := 0; j < 3 && left + j <= right; j++ {
            strs = append(strs, products[left + j])
        }
        
        result = append(result, strs)
    }
    
    return result
    
}

File: 1268-search-suggestions-system/1268-search-suggestions-system.py
class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        
        products.sort()
        l, r = 0, len(products) - 1
        res = []
        for i in range(len(searchWord)):
            c = searchWord[i]
            while l <= r and ( len(products[l]) <= i or products[l][i] != c ):
                l += 1
            while l <= r and ( len(products[r]) <= i or products[r][i] != c):
                r -= 1
            res.append([])
            for j in range(min(3, r - l + 1)):
                res[-1].append(products[l + j])
        return res